File: pandemic-game/src/City.js
import './App.css';

export default function City(props) {

    const cubes = Object.entries(props.city.cubes).map((cube) => {
        if(cube[1] > 0) {
            return <div className={`cube ${cube[0]}`}>{cube[1]}</div>
        }
    });

    let researchStation;
    if (props.city.has_research_station) {
        researchStation = <div className="researchStation"></div>
    }

    return (
        <div id={props.city.name} className={`city ${props.city.color}`}>
            {cubes}
            {researchStation}
            <div className="cityName">{props.city.visible_name}</div>
        </div>)
}

File: pandemic/app/game/city.rb
class City
  attr_accessor :name, :color, :connecting_cities, :cubes, :has_research_station

  def initialize(name, color)
    @name = name
    @visible_name = name.gsub("_", " ").titleize
    @color = color
    @cubes = {"blue" => 0, "yellow" => 0, "black" => 0, "red" => 0}
    @has_research_station = false
  end

  def add_cube
    add_colored_cube(@color)
  end

  def add_colored_cube(color)
    current_cubes = @cubes[color]
    if (current_cubes == 3)
      return true
    else
      @cubes[color] = current_cubes + 1
      return false
    end
  end

  def add_epidemic_cubes
    previous_count = @cubes[@color]
    @cubes[@color] = 3
    return 3 - previous_count
  end
end